docs/website/docs/reference/bindings/c-api.md

```diff
@@ -1,216 +1,353 @@
+
+#### High level - Session
+
+!!! todo - "Under construction, more coming soon"
+
+#### High level - Instance
+
+!!! todo - "Under construction, more coming soon"
+
+#### High level - Call
+
+!!! todo - "Under construction, more coming soon"
+
+#### Low level - VM
+
+<!-- TODO(scotttodd): VM module diagram (bytecode, HAL, custom) -->
+
+By default, IREE uses its own tiny Virtual Machine (VM) at runtime to interpret
+program instructions on the host system. VM instructions may also be lowered
+further to LLVM IR, C, or other representations for static or resource
+constrained deployment.
+
+The VM supports generic operations like loads, stores, arithmetic, function
+calls, and control flow. The VM builds streams of more complex program logic and
+dense math into HAL command buffers that are dispatched to hardware backends.
+
+* VM _instances_ can serve multiple isolated execution _contexts_
+* VM _contexts_ are effectively sandboxes for loading modules and running
+  programs
+* VM _modules_ provide extra functionality to execution _contexts_, such as
+  access to hardware accelerators through the HAL. Compiled user programs are
+  also modules.
+
+#### Low level - HAL
+
+<!-- TODO(scotttodd): command buffer construction -> dispatch diagram -->
+<!-- TODO(scotttodd): input buffers -> output buffers diagram -->
+<!-- TODO(scotttodd): HAL interface diagram -->
+
+* HAL _drivers_ are used to enumerate and create HAL _devices_
+* HAL _devices_ interface with hardware, such as by allocating device memory,
+  preparing executables, recording and dispatching command buffers, and
+  synchronizing with the host
+* HAL _buffers_ and _buffer views_ represent storage and shaped/typed views
+  into that storage (aka "tensors")
 
-<!-- ## Advanced usage -->
+### Usage
 
-<!-- TODO(scotttodd): async execution and synchronization -->
-<!-- TODO(scotttodd): specialized HAL APIs -->
-<!-- TODO(scotttodd): threadpools -->
-<!-- TODO(scotttodd): heterogenous execution -->
+!!! todo - "Under construction, more coming soon"
 
-<!-- ## Troubleshooting -->
+#### Samples
 
-<!-- TODO(scotttodd): link to GitHub issues -->
-<!-- TODO(scotttodd): compiler/runtime compatibility -->
-<!-- TODO(scotttodd): common problems? object ownership? loaded modules (HAL)? -->
+| Source location | Description |
+| --------------- | ----------- |
+[iree `runtime/demo/`](https://github.com/openxla/iree/blob/main/runtime/src/iree/runtime/demo/) | In-tree demos of the high level runtime API
+[iree `samples/`](https://github.com/openxla/iree/tree/main/samples) | In-tree sample applications
+[iree-template-runtime-cmake](https://github.com/benvanik/iree-template-runtime-cmake/) | Template repository for runtime applications
+[iree-template-cpp](https://github.com/iml130/iree-template-cpp) | Demonstration of integration into a downstream project
+[iree-samples `runtime-library/`](https://github.com/iree-org/iree-samples/tree/main/runtime-library) | Shared runtime library builder
```

Comment:
I'm in the process of refreshing `iree-template-runtime-cmake` and moving it to iree-org.

Could link to https://github.com/iree-org/iree-samples/tree/main/cpp ... and maybe move https://github.com/iree-org/iree-samples/tree/main/runtime-library into that directory? (I was imagining a `python/` folder next to `cpp/`, but the `iree-samples` repo is kind of a mess...)